```{r}
## Load libraries
library(devtools)
library(rmarkdown)
library(dada2)
library(ShortRead)
library(ggplot2)
library("ggforce")
library(phyloseq)
library(vegan)
library(knitr)
library(ALDEx2)
library(CoDaSeq)
library(zCompositions)
library(igraph)
library(car)
library(grDevices)
library(propr)
library(cowplot)
library(randomcoloR)
library(dplyr)
library(reshape2)
library(tibble)
library(exactRankTests)
library(nlme)
library(data.table)
library(Rmisc)
##set wd
setwd("~/Desktop/PANAMA_HYPOXIA")
## Save version information
writeLines(capture.output(sessionInfo()), "sessionInfo.txt")
```


```{r}
###### Quality-filter reads and create Amplicon Sequence Variant tables
# put parsed, adaptors & primers removed, unjoined (R1 and R2 separate) fastq files
# into directory for DADA2 & make sure the full path is updated in the next line:
###adjust path name
path <- "~/Desktop/PANAMA_HYPOXIA/cutadapt"
list.files(path)
```


```{r}
# Forward and reverse fastq filenames have format: SAMPLENAME_R1_cut.fastq.gz and SAMPLENAME_R2_cut.fastq.gz
# Samplename is everything before the first underscore
fnFs <- sort(list.files(path, pattern="_R1_cut.fastq.gz", full.names = TRUE))
fnRs <- sort(list.files(path, pattern="_R2_cut.fastq.gz", full.names = TRUE))
# Extract sample names, assuming filenames have format: SAMPLENAME_XXX.fastq
sample.names <- sapply(strsplit(basename(fnFs), "_"), `[`, 1)
```


```{r}
# Examine quality profiles of forward and reverse reads
plotQualityProfile(fnFs[1:6])
plotQualityProfile(fnRs[1:6])
```


```{r}
# Perform filtering and trimming
# Assign the filenames for the filtered fastq.gz files.
# Make directory and filenames for the filtered fastqs
filt_path <- file.path(path, "filtered") # Place filtered files in filtered/ subdirectory
filtFs <- file.path(filt_path, paste0(sample.names, "_F_filt.fastq.gz"))
filtRs <- file.path(filt_path, paste0(sample.names, "_R_filt.fastq.gz"))
```


```{r}
# Filter the forward and reverse reads
out <- filterAndTrim(fnFs, filtFs, fnRs, filtRs, truncLen=c(150,150),
                     maxN=0, maxEE=c(2,2), truncQ=2, rm.phix=TRUE,
                     compress=TRUE, multithread=TRUE) # On Windows set multithread=FALSE
head(out)
```


```{r}
# Learn the Error Rates, it TAKES TIME!
# Forward reads
errF <- learnErrors(filtFs, multithread=TRUE)
# Reverse reads
errR <- learnErrors(filtRs, multithread=TRUE)
# visualize the estimated error rates
plotErrors(errF, nominalQ=TRUE)
```


```{r}
# Dereplicate the filtered fastq files
derepFs <- derepFastq(filtFs, verbose=TRUE)
derepRs <- derepFastq(filtRs, verbose=TRUE)
# Name the derep-class objects by the sample names
names(derepFs) <- sample.names
names(derepRs) <- sample.names
# Infer the sequence variants in each sample
dadaFs <- dada(derepFs, err=errF, multithread=TRUE)
dadaRs <- dada(derepRs, err=errR, multithread=TRUE)
# Inspecting the dada-class object returned by dada:
dadaFs[[1]]
# Merge the denoised forward and reverse reads:
mergers <- mergePairs(dadaFs, derepFs, dadaRs, derepRs, verbose=TRUE)
# Inspect the merger data.frame from the first sample
head(mergers[[1]])
# Construct sequence table
seqtab <- makeSequenceTable(mergers) ## The sequences being tabled vary in length.
dim(seqtab)
# Inspect distribution of sequence lengths
table(nchar(getSequences(seqtab)))
#Remove chimeric sequences:
seqtab.nochim <- removeBimeraDenovo(seqtab, method="consensus", multithread=TRUE, verbose=TRUE)
dim(seqtab.nochim)
sum(seqtab.nochim)/sum(seqtab)
```


```{r}
# Track reads through the pipeline
# As a final check of our progress, we’ll look at the number of reads that made it through each step in the pipeline
getN <- function(x) sum(getUniques(x))
track <- cbind(out, sapply(dadaFs, getN), sapply(mergers, getN), rowSums(seqtab), rowSums(seqtab.nochim))
# If processing a single sample, remove the sapply calls: e.g. replace sapply(dadaFs, getN) with getN(dadaFs)
colnames(track) <- c("input", "filtered", "denoised", "merged", "tabled", "nonchim")
rownames(track) <- sample.names
head(track)
write.table(track, "dada_read_stats.txt",sep="\t",col.names=NA)
# SAVE THIS FILE SO YOU DON'T HAVE TO REPEAT ALL OF THE ABOVE STEPS, adjust name
saveRDS(seqtab.nochim, file="~/Desktop/PANAMA_HYPOXIA/seqtab.nochim.rds")
# RELOAD THE SAVED INFO FROM HERE (if you have closed the project):
# seqtab.nochim <- readRDS("~/Desktop/PANAMA_HYPOXIA/seqtab.nochim.rds")
seqtab.nochim <- readRDS("~/Desktop/PANAMA_HYPOXIA/seqtab.nochim.rds")
```


```{r}
# Assign taxonomy
# Make sure the appropriate database is available in the DADA2 directory
taxa <- assignTaxonomy(seqtab.nochim, "~/Desktop/PANAMA_HYPOXIA/silva_nr_v132_train_set.fa.gz", multithread=TRUE)
# FIX the NAs in the taxa table
taxon <- as.data.frame(taxa,stringsAsFactors=FALSE)
taxon$Phylum[is.na(taxon$Phylum)] <- taxon$Kingdom[is.na(taxon$Phylum)]
taxon$Class[is.na(taxon$Class)] <- taxon$Phylum[is.na(taxon$Class)]
taxon$Order[is.na(taxon$Order)] <- taxon$Class[is.na(taxon$Order)]
taxon$Family[is.na(taxon$Family)] <- taxon$Order[is.na(taxon$Family)]
taxon$Genus[is.na(taxon$Genus)] <- taxon$Family[is.na(taxon$Genus)]
write.table(taxon,"Panama_silva_taxa_table.txt",sep="\t",col.names=NA)
write.table(seqtab.nochim, "Panama_silva_otu_table.txt",sep="\t",col.names=NA)
```


```{r}
# Create phyloseq object from otu and taxonomy tables from dada2, along with the sample metadata.
setwd("~/Desktop/PANAMA_HYPOXIA")

otu <- read.table("Panama_silva_otu_table.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("Panama_silva_taxa_table.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("Panama_metadata.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps
#24084 taxa and 57 samples
```


```{r}
# remove chloroplasts and mitochondria and Eukaryota
get_taxa_unique(ps, "Family") #679
get_taxa_unique(ps, "Order") #411
get_taxa_unique(ps, "Kingdom") #2
ps <- subset_taxa(ps, Family !="Mitochondria")
ps <- subset_taxa(ps, Order !="Chloroplast")
ps <- subset_taxa(ps, Kingdom !="Eukaryota")
ps <- subset_taxa(ps, Kingdom !="NA")
get_taxa_unique(ps, "Family") #675
get_taxa_unique(ps, "Order") #408
get_taxa_unique(ps, "Kingdom") #2
ps
#22823 taxa and 57 samples
# filtered taxa with phyloseq, now export cleaned otu and taxa tables from phyloseq
otu = as(otu_table(ps), "matrix")
taxon = as(tax_table(ps), "matrix")
metadata = as(sample_data(ps), "matrix")
write.table(otu,"Panama_silva_nochloronomito_otu_table.txt",sep="\t",col.names=NA)
write.table(taxon,"Panama_silva_nochloronomito_taxa_table.txt",sep="\t",col.names=NA)
# remove control samples for plotting, remaining samples = 56
ps = subset_samples(ps, Site != "control") ##! is how you exclude a sample
ps
#22823 taxa and 56 samples
```

```{r}
##Subset to only include samples from Punta Caracol for analysis
ps_pc = subset_samples(ps, Site == "Punta Caracol")
```

```{r}
# plot number of observed ASVs in coral samples
plot_richness(ps_pc,x="Treatment",color="Species",measures=c("Observed"))
# look at data and chose filtering method for very low abundance ASVs
ntaxa(ps_pc) #22823
ps5<-filter_taxa(ps_pc, function(x) mean(x) >5, TRUE)
ntaxa(ps5) #920
get_taxa_unique(ps_pc, "Genus") # 1000
get_taxa_unique(ps5, "Genus") #299
# filtered ASVs with very low abundance with phyloseq, now export otu and taxa tables from phyloseq for codaseq
otu = as(otu_table(ps5), "matrix")
taxon = as(tax_table(ps5), "matrix")
metadata = as(sample_data(ps5), "matrix")
write.table(otu,"Panama_ps5_silva_nochloronomito_otu_table.txt",sep="\t",col.names=NA)
write.table(taxon,"Panama_ps5_silva_nochloronomito_taxa_table.txt",sep="\t",col.names=NA)
write.table(metadata,"Panama_ps5_silva_metadata.txt",sep="\t",col.names=NA)
```


```{r}
######### Perform center-log-ratio transformation on ASVs and calculate Aitchison Distance and principal components
# READ IN OTU data that has been filtered for very low abundance sequences; do not clear data here. Keep phyloseq object ps5 for anosim/permanova
otu <- read.table("Panama_ps5_silva_nochloronomito_otu_table.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("Panama_ps5_silva_nochloronomito_taxa_table.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("Panama_ps5_silva_metadata.txt",sep="\t",header=T,row.names=1)
genus<-as.character(taxon$Genus)
# First, replace 0 values with an estimate (because normalization is taking log, can't have 0)
# Also transposing here, need samples as rows
d.czm <- cmultRepl(t(otu), method="CZM", label=0)
# Perform the center-log-CLR) transformation 
d.clr <- codaSeq.clr(d.czm)
# transpose matrix of CLR transformed data for ordination and dendrogram
E.clr <- t(d.clr)
# plot compositional PCA biplot (perform a singular value decomposition)
d.pcx <- prcomp(E.clr)
# calculate percent variance explained for the axis labels
pc1 <- round(d.pcx$sdev[1]^2/sum(d.pcx$sdev^2),2)
pc2 <- round(d.pcx$sdev[2]^2/sum(d.pcx$sdev^2),2)
xlab <- paste("PC1: ", pc1, sep="")
ylab <- paste("PC2: ", pc2, sep="")
biplot(d.pcx, cex=c(0.6,0.4), var.axes=F,scale=1, xlab=xlab, ylab=ylab, ylabs=genus)
summary(d.pcx)
str(d.pcx)
screeplot(d.pcx)

# replot PCA with ggplot2 (showing samples only), PCA PLOT, SPECIES AND TREATMENT COMPARISON
df_out <- as.data.frame(d.pcx$x)
theme_set(theme_bw()+theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank()))
cols<-c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
#pdf("PCA_species_treatment.pdf",width=8.5)
p<-ggplot(df_out,aes(x=PC1,y=PC2,fill=samples$Species,shape=samples$Treatment))
p<-p+geom_point(size=3)+
  theme(axis.title = element_text(size=14))+
  theme(axis.text=element_text(size=12))+
  theme(legend.title = element_text(size=14))+
  theme(legend.text = element_text(size=12))+
  scale_shape_manual(values=c(21,22,24))+
  guides(fill = guide_legend(override.aes=list(shape=21)))
p + labs(x=xlab, y=ylab, fill="Species", shape="Treatment") + coord_fixed()
#dev.off()
```

```{r}
##### Use phyloseq/vegan to perform ANOSIM/PERMANOVA
# set metadata as factors for anosim
treat<-as.character(samples$Treatment)
species<-as.character(samples$Species)
origin<-as.character(samples$Origin)
site<-as.character(samples$Site)
logger<-as.character(samples$Logger.MiniDOT)
# set Aitchison distance as the distance measure for the anosim
dist.clr <- dist(E.clr)
# anosim between groups using Aitchison distance, ARE THE GROUPS DIFFERENT (look at between data)
ano.treat <- anosim(dist.clr, treat, permutations=999)
#pdf("Panama_TREATMENT_EFFECT_ANOSIM.pdf",width=8.5)
plot(ano.treat)
#dev.off()
####### Use phyloseq/vegan to perform ANOSIM/PERMANOVA
#Species
ano.species <- anosim(dist.clr, species, permutations=999)
#pdf("Panama_ANOSIM_Species.pdf",width=8.5)
plot(ano.species)
#dev.off()
#Origin
ano.origin <- anosim(dist.clr, origin, permutations=999)
#pdf("Panama_ANOSIM_origin.pdf",width=8.5)
plot(ano.origin)
#dev.off()
#Logger
ano.logger <- anosim(dist.clr, logger, permutations=999)
#pdf("Panama_ANOSIM_logger.pdf",width=8.5)
plot(ano.logger)
#dev.off()

#permanova, treatment and species
perm<-adonis2(dist.clr~treat*species,as(sample_data(ps_pc),"data.frame"))
print(perm)
```

```{r}
##Use ANOSIM to determine differences between Finca and Tierra Oscura corals
#subset to exclode Punta Caracol samples
ps_fto<-subset_samples(ps, Species !="Agaricia lamarcki")
ps_fto<-subset_samples(ps_fto, Site !="Punta Caracol")

# plot number of observed ASVs in coral samples
plot_richness(ps_fto,x="Treatment",color="Species",measures=c("Observed"))
# look at data and chose filtering method for very low abundance ASVs
ntaxa(ps_fto) #22823
ps_fto_5<-filter_taxa(ps_fto, function(x) mean(x) >5, TRUE)
ntaxa(ps_fto_5) #569
get_taxa_unique(ps_fto, "Genus") # 1000
get_taxa_unique(ps_fto_5, "Genus") #232
# filtered ASVs with very low abundance with phyloseq, now export otu and taxa tables from phyloseq for codaseq
otu = as(otu_table(ps_fto_5), "matrix")
taxon = as(tax_table(ps_fto_5), "matrix")
metadata = as(sample_data(ps_fto_5), "matrix")
write.table(otu,"Panama_psfto5_silva_nochloronomito_otu_table.txt",sep="\t",col.names=NA)
write.table(taxon,"Panama_psfto5_silva_nochloronomito_taxa_table.txt",sep="\t",col.names=NA)
write.table(metadata,"Panama_psfto5_silva_metadata.txt",sep="\t",col.names=NA)

######### Perform center-log-ratio transformation on ASVs and calculate Aitchison Distance and principal components
# READ IN OTU data that has been filtered for very low abundance sequences; do not clear data here. Keep phyloseq object psfto5 for anosim/permanova
otu <- read.table("Panama_psfto5_silva_nochloronomito_otu_table.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("Panama_psfto5_silva_nochloronomito_taxa_table.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("Panama_psfto5_silva_metadata.txt",sep="\t",header=T,row.names=1)
genus<-as.character(taxon$Genus)
# First, replace 0 values with an estimate (because normalization is taking log, can't have 0)
# Also transposing here, need samples as rows
d.czm <- cmultRepl(t(otu), method="CZM", label=0)
# Perform the center-log-ratio (CLR) transformation 
d.clr <- codaSeq.clr(d.czm)
# transpose matrix of CLR transformed data for ordination and dendrogram
E.clr <- t(d.clr)

site<-as.character(samples$Site)
# set Aitchison distance as the distance measure for the anosim
dist.clr <- dist(E.clr)
ano.site <- anosim(dist.clr, site, permutations=999)
#pdf("Panama_SITE_EFFECT_ANOSIM.pdf",width=8.5)
plot(ano.site)
#dev.off()
```


```{r}
##linear regression 
df_pc1<-subset(df_out, select=c(PC1))
write.table(df_pc1,"PC1_values.txt",sep="\t",col.names=NA)

##added O2 concentrations to table. Load table back in.
regression<-read.table("PC1_O2_metadata.txt",sep="\t",header=TRUE, row.names=1)

#regression
x<-c(regression$DO)
y<-c(regression$PC1)
model<-lm(x~y)
summary(model)

##graph
cols<-c("Control plots"="#0072B2","Hypoxia plots"="#56B4E9")

pdf("LinearRegression.pdf",width=8)
theme_set(theme_bw()+theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank()))
p<-ggplot(regression, aes(x=x, y=y))+
  geom_point(size=3,aes(color=Treatment),stat="identity")+
  scale_color_manual(values=cols)+
  theme(axis.title = element_text(size=16))+
  theme(axis.text=element_text(size=16))+
  theme(legend.title = element_text(size=16))+
  theme(legend.text = element_text(size=14))+
  ylab("PC1") + xlab("Dissolved oxygen (mg/L)")

# Add the linear regression line to the plotted data
p <- p + geom_smooth(method="lm", col="black")
p
dev.off()
```


```{r}
############ Alpha Diversity #############
##############################################
#Alpha Diversity - number of unique OTUS (Observed) or taking into account the proportion of OTUs across samples (Shannon)

pdf("Alpha_diversity",width=10, height=6)
p<-plot_richness(ps_pc, x="Treatment",color="Species", measures=c("Chao1", "Shannon", "Simpson"))
p<-p+geom_point(size=3)+
  theme(axis.title = element_text(size=12))+
  theme(axis.text=element_text(size=10))+
  theme(axis.text.x = element_text(angle = 45, hjust=.5))+
  theme(legend.title = element_text(size=12))+
  theme(legend.text = element_text(size=10))
p
dev.off()

```


```{r}
##########Beta Diversity Dispersions
#calculate multivariate dispersions based on treatment
mod <-betadisper(dist.clr, treat)
#one way anova
anova(mod)
#boxplots
plot(mod)
boxplot(mod)

## Compute mean distance to centroid per group
#this just prints values on the console
tapply(mod$distances, treat, mean)
## Same, but variance instead
tapply(mod$distances, treat, var)

#Get the distances to centroid from the model
mod$distances
dis <- mod$distances
#melt
dis.melt <- melt(dis)
#move rownames to columns so we can merge the dispersion values and metadata
dis.melt$Sample <- rownames(dis.melt)
samples$Sample <- rownames(samples)
#merge metadata and dispersion 
dis.treat <- merge(samples, dis.melt)
#rename column
colnames(dis.treat)[7] <- "distance"

#run linear model to test significance
distlm <-lm(distance~Species*Treatment, data=dis.treat)
summary(distlm)
anova(distlm)

# plot average dispersion by group, with all points shown    
cols<-c("Siderastrea siderea"="#55BCC2","Agaricia lamarcki"="#E77D72")
dis.treat$Treatment<-factor(dis.treat$Treatment, levels=c("Control plots","Hypoxia plots"))
theme_set(theme_bw()+theme(panel.grid.major = element_blank(),panel.grid.minor = element_blank()))
pdf("DistanceToCentroid.pdf",width=8.5)
p2<-ggplot(dis.treat,aes(x=Treatment,y=distance))+
  geom_boxplot()+
  geom_jitter(position=position_jitter(width=.1, height=0),aes(color=Species),size=3)+
  scale_color_manual(values=cols)+
  theme(text=element_text(size=16))+
  theme(legend.title = element_text(size=14))+
  theme(legend.text = element_text(size=12))+
  ylab("Distance to Centroid") + xlab("Treatment")
p2 
dev.off()
```


```{r}
########## BARCHARTS
# READ IN OTU data that has been filtered for very low abundance sequences. 
setwd("~/Desktop/PANAMA_HYPOXIA")
##make a phyloseq object to be manipulated
otu <- read.table("Panama_ps5_silva_nochloronomito_otu_table.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("Panama_ps5_silva_nochloronomito_taxa_table.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("Panama_ps5_silva_metadata.txt",sep="\t",header=T,row.names=1)
genus<-as.character(taxon$Genus)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps5 <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps5
ps_ra<-transform_sample_counts(ps5, function(OTU) OTU/sum(OTU))
ps_ra_alam = subset_samples(ps_ra, Species == "Agaricia lamarcki")
ps_ra_ssid = subset_samples(ps_ra, Species == "Siderastrea siderea")
#figure out how many colors you need
get_taxa_unique(ps_ra, "Order") #106
get_taxa_unique(ps_ra, "Class") #38
#you can make n any number of colors you want; with as much difference between the colors as possible (distinct colors)
n <- 38
palette <- distinctColorPalette(n)  #you can rerun this line to get a new selection of colors
#save list of colors for palette that you like, this kind of works:
cat(capture.output(print(palette), file="palette38.txt"))
palette38<-c("#AC99E4","#BA4AB4","#4BA68D","#A240ED","#79DE65","#E7A24F","#A3A47A","#6EE295","#6AEB44","#CEF146","#D9CA79","#E2B7E9","#E0E8DB","#D580B8","#9DE9C1","#6DA0E2","#95B2AA","#6CE8ED","#E25D60","#DAEAB6","#7DAA5E","#CEEB86","#7080E4","#B4E7E3","#E840DB","#AE98B3","#E8D2DD","#BACCEC","#E8D84A","#B0735D","#53EECF","#664D75","#E5559B","#E8BD9F","#E297A4","#6D56CF","#DF87EB","#69B8D9")
##put treatments in a different order
sample_data(ps_ra_alam)$Treatment<-factor(sample_data(ps_ra_alam)$Treatment,levels=c("Control plots","Hypoxia plots"))
p1=plot_bar(ps_ra_alam, fill="Class")+
  geom_bar(aes(fill=Class), stat="identity",position="stack")+theme(strip.text=element_text(face="bold"))+
  theme(axis.text.x = element_blank(),axis.ticks = element_blank())+scale_fill_manual(values=palette38)+
  facet_grid(.~Treatment,scales="free",space="free")+
  ggtitle("Agaricia lamarcki")+
  theme(plot.title = element_text(face="italic"))+
  theme(legend.position = "bottom")+
  theme(axis.title.x = element_blank())
p1
##put treatments in a different order
sample_data(ps_ra_ssid)$Treatment<-factor(sample_data(ps_ra_ssid)$Treatment,levels=c("Control plots","Hypoxia plots"))
p2=plot_bar(ps_ra_ssid, fill="Class")+
  geom_bar(aes(fill=Class), stat="identity",position="stack")+theme(strip.text=element_text(face="bold"))+
  theme(axis.text.x = element_blank(),axis.ticks = element_blank())+scale_fill_manual(values=palette38)+
  facet_grid(.~Treatment,scales="free",space="free")+
  ggtitle("Siderastrea siderea")+
  theme(plot.title = element_text(face="italic"))+
  theme(legend.position="bottom")+
  theme(axis.title.x = element_blank())
p2
# adjust width and height until it looks right for double columns
pdf("Panama_Barcharts_Class.pdf",width=10, height=10)
plot_grid(p1,p2,labels=c("A","B"), ncol=1, nrow=2)
dev.off()
```


```{r}
################################# ANCOM TEST OF DIFFERENTIALLY ABUNDANT FAMILIES
#ANCOM Function - compare across multiple treatments groups using a compositional appproach
#https://sites.google.com/site/siddharthamandal1985/research
###Need to run this first in order to run ANCOM on data
ancom.W = function(otu_data,var_data,
                   adjusted,repeated,
                   main.var,adj.formula,
                   repeat.var,long,rand.formula,
                   multcorr,sig){
  
  n_otu=dim(otu_data)[2]-1
  
  otu_ids=colnames(otu_data)[-1]
  
  if(repeated==F){
    data_comp=data.frame(merge(otu_data,var_data,by="Sample.ID",all.y=T),row.names=NULL)
    #data_comp=data.frame(merge(otu_data,var_data[,c("Sample.ID",main.var)],by="Sample.ID",all.y=T),row.names=NULL)
  }else if(repeated==T){
    data_comp=data.frame(merge(otu_data,var_data,by="Sample.ID"),row.names=NULL)
    # data_comp=data.frame(merge(otu_data,var_data[,c("Sample.ID",main.var,repeat.var)],by="Sample.ID"),row.names=NULL)
  }
  
  base.formula = paste0("lr ~ ",main.var)
  if(repeated==T){
    repeat.formula = paste0(base.formula," | ", repeat.var)
  }
  if(adjusted==T){
    adjusted.formula = paste0(base.formula," + ", adj.formula)
  }
  
  if( adjusted == F & repeated == F ){
    fformula  <- formula(base.formula)
  } else if( adjusted == F & repeated == T & long == T ){
    fformula  <- formula(base.formula)   
  }else if( adjusted == F & repeated == T & long == F ){
    fformula  <- formula(repeat.formula)   
  }else if( adjusted == T & repeated == F  ){
    fformula  <- formula(adjusted.formula)   
  }else if( adjusted == T & repeated == T  ){
    fformula  <- formula(adjusted.formula)   
  }else{
    stop("Problem with data. Dataset should contain OTU abundances, groups, 
         and optionally an ID for repeated measures.")
  }
  
  
  
  if( repeated==FALSE & adjusted == FALSE){
    if( length(unique(data_comp[,which(colnames(data_comp)==main.var)]))==2 ){
      tfun <- exactRankTests::wilcox.exact
    } else{
      tfun <- stats::kruskal.test
    }
  }else if( repeated==FALSE & adjusted == TRUE){
    tfun <- stats::aov
  }else if( repeated== TRUE & adjusted == FALSE & long == FALSE){
    tfun <- stats::friedman.test
  }else if( repeated== TRUE & adjusted == FALSE & long == TRUE){
    tfun <- nlme::lme
  }else if( repeated== TRUE & adjusted == TRUE){
    tfun <- nlme::lme
  }
  
  logratio.mat <- matrix(NA, nrow=n_otu, ncol=n_otu)
  for(ii in 1:(n_otu-1)){
    for(jj in (ii+1):n_otu){
      data.pair <- data_comp[,which(colnames(data_comp)%in%otu_ids[c(ii,jj)])]
      lr <- log((1+as.numeric(data.pair[,1]))/(1+as.numeric(data.pair[,2])))
      
      lr_dat <- data.frame( lr=lr, data_comp,row.names=NULL )
      
      if(adjusted==FALSE&repeated==FALSE){  ## Wilcox, Kruskal Wallis
        logratio.mat[ii,jj] <- tfun( formula=fformula, data = lr_dat)$p.value
      }else if(adjusted==FALSE&repeated==TRUE&long==FALSE){ ## Friedman's 
        logratio.mat[ii,jj] <- tfun( formula=fformula, data = lr_dat)$p.value
      }else if(adjusted==TRUE&repeated==FALSE){ ## ANOVA
        model=tfun(formula=fformula, data = lr_dat,na.action=na.omit)   
        picker=which(gsub(" ","",row.names(summary(model)[[1]]))==main.var)  
        logratio.mat[ii,jj] <- summary(model)[[1]][["Pr(>F)"]][picker]
      }else if(repeated==TRUE&long==TRUE){ ## GEE
        model=tfun(fixed=fformula,data = lr_dat,
                   random = formula(rand.formula),
                   correlation=corAR1(),
                   na.action=na.omit)   
        picker=which(gsub(" ","",row.names(anova(model)))==main.var)
        logratio.mat[ii,jj] <- anova(model)[["p-value"]][picker]
      }
      
    }
  } 
  
  ind <- lower.tri(logratio.mat)
  logratio.mat[ind] <- t(logratio.mat)[ind]
  
  
  logratio.mat[which(is.finite(logratio.mat)==FALSE)] <- 1
  
  mc.pval <- t(apply(logratio.mat,1,function(x){
    s <- p.adjust(x, method = "BH")
    return(s)
  }))
  
  a <- logratio.mat[upper.tri(logratio.mat,diag=FALSE)==TRUE]
  
  b <- matrix(0,ncol=n_otu,nrow=n_otu)
  b[upper.tri(b)==T] <- p.adjust(a, method = "BH")
  diag(b)  <- NA
  ind.1    <- lower.tri(b)
  b[ind.1] <- t(b)[ind.1]
  
  #########################################
  ### Code to extract surrogate p-value
  surr.pval <- apply(mc.pval,1,function(x){
    s0=quantile(x[which(as.numeric(as.character(x))<sig)],0.95)
    # s0=max(x[which(as.numeric(as.character(x))<alpha)])
    return(s0)
  })
  #########################################
  ### Conservative
  if(multcorr==1){
    W <- apply(b,1,function(x){
      subp <- length(which(x<sig))
    })
    ### Moderate
  } else if(multcorr==2){
    W <- apply(mc.pval,1,function(x){
      subp <- length(which(x<sig))
    })
    ### No correction
  } else if(multcorr==3){
    W <- apply(logratio.mat,1,function(x){
      subp <- length(which(x<sig))
    })
  }
  
  return(W)
}
ANCOM.main = function(OTUdat,Vardat,
                      adjusted,repeated,
                      main.var,adj.formula,
                      repeat.var,longitudinal,
                      random.formula,
                      multcorr,sig,
                      prev.cut){
  p.zeroes=apply(OTUdat[,-1],2,function(x){
    s=length(which(x==0))/length(x)
  })
  OTUdat.thinned=OTUdat[,c(1,1+which(p.zeroes<prev.cut))]
  otu.names=colnames(OTUdat.thinned)[-1]
  W.detected   <- ancom.W(OTUdat.thinned,Vardat,
                          adjusted,repeated,
                          main.var,adj.formula,
                          repeat.var,longitudinal,random.formula,
                          multcorr,sig)
  W_stat       <- W.detected
  W_frame = data.frame(otu.names,W_stat,row.names=NULL)
  W_frame = W_frame[order(-W_frame$W_stat),]
  W_frame$detected_0.9=rep(FALSE,dim(W_frame)[1])
  W_frame$detected_0.8=rep(FALSE,dim(W_frame)[1])
  W_frame$detected_0.7=rep(FALSE,dim(W_frame)[1])
  W_frame$detected_0.6=rep(FALSE,dim(W_frame)[1])
  W_frame$detected_0.9[which(W_frame$W_stat>0.9*(dim(OTUdat.thinned[,-1])[2]-1))]=TRUE
  W_frame$detected_0.8[which(W_frame$W_stat>0.8*(dim(OTUdat.thinned[,-1])[2]-1))]=TRUE
  W_frame$detected_0.7[which(W_frame$W_stat>0.7*(dim(OTUdat.thinned[,-1])[2]-1))]=TRUE
  W_frame$detected_0.6[which(W_frame$W_stat>0.6*(dim(OTUdat.thinned[,-1])[2]-1))]=TRUE
  final_results=list(W_frame)
  names(final_results)=c("W.taxa")
  return(final_results)
}
```


```{r}
#####ANCOM results
otu <- read.table("Panama_silva_nochloronomito_otu_table.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("Panama_silva_nochloronomito_taxa_table.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("Panama_metadata.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps
#22823 taxa and 57 samples
# remove control samples for plotting, remaining samples = 56
ps = subset_samples(ps, Species != "control")
ps

get_taxa_unique(ps, "Family")

#Separate coral species
ps_ssid <- subset_samples(ps, Species == "Siderastrea siderea")
ps_alam <- subset_samples(ps, Species == "Agaricia lamarcki")

############Siderastrea siderea
# aggregate ASVs by family
dat <- tax_glom(ps_ssid, taxrank = "Family")

#melt the data, so it's like a dataframe
datm <- psmelt(dat)

#Cast the new datatable with columns that are of interest
datc <- data.table::dcast(datm, Sample + Logger.MiniDOT + Treatment + Origin + Species ~ Family, value.var = 'Abundance', fun.aggregate = sum)

dim(datc) #dimensions of the table
#680 families

otud <- datc[,c(1,7:680)] #select the first column, and then all of the taxa columns  
colnames(otud)[1] <- "Sample.ID" #rename the first column to Sample.ID - this is to match ANCOM syntax
metadat <- sample_data(ps_ssid) #get the sample data
metadat <- as.data.frame(as.matrix(metadat)) #make into into a matrix
# at this point, my sample id numbers are the row names, not a separate column, move row names to column with dplylr
metadat <- tibble::rownames_to_column(metadat, "Sample.ID") #make sure the sample names column is called Sample.ID
names(otud) <- make.names(names(otud)) #get the names from the table for 
otu_test <- otud #rename otud to otu_test, for syntax in ANCOM
metadat <- select(metadat, c("Sample.ID","Species","Treatment","Origin")) # use select to only use treatment columns of interest
map_test <- metadat #rename map_TEst
Vardat <- map_test #specify that this for Vardat - ANCOM syntax
#### ANCOM test - not adjusted, more than 2 levels = Kruskal Wallis
comparison_test_treat=ANCOM.main(OTUdat=otu_test, #calling the OTU table
                                 Vardat=map_test, #calling the metadata
                                 adjusted=FALSE, #true if covariates are to be included for adjustment
                                 repeated=FALSE, #repeated measure
                                 main.var="Treatment", #main variable or fator
                                 adj.formula= NULL, #other factors to include
                                 repeat.var=FALSE, #repeated measure
                                 long = FALSE, #longitudinal study
                                 multcorr=2,
                                 sig=0.05, #significance level
                                 prev.cut=0.90) #ASV needs to be in more the 10% of the samples #OTUs with proportion of zeroes greater than prev.cut are not included in the analysis
res <- comparison_test_treat$W.taxa #taxa that significantly vary across factor level of interest
write.table(res,"SSID_ANCOM_family_KruskallWallis_Treatment.txt",sep="\t",col.names=NA)
res2 <- res[which(res$detected_0.7==TRUE),] #0.7 is the proportion of samples its detected in
sig_sites <- glue::glue_collapse(droplevels(factor(res2$otu.names)), sep = ", ") #this is to get a list of the families that are different
print(sig_sites)
#Family_XII, Marinifilaceae, Pseudomonadaceae, Nitrincolaceae, Marinobacteraceae, Parvibaculaceae, Shewanellaceae, Halomonadaceae, Francisellaceae

#Calculate relative abundance
datc_relabund <-  sweep(datc[,7:680], 1, rowSums(datc[,7:680]), '/')
datc_relnames <- cbind(datc[,1:6],datc_relabund)

#only selet the significant families
sig_dis <- select(datc_relnames, Sample, Species, Treatment, Family_XII, Marinifilaceae, Pseudomonadaceae, Nitrincolaceae, Marinobacteraceae, Parvibaculaceae, Shewanellaceae, Halomonadaceae, Francisellaceae)
sig_long <- melt(sig_dis, id.vars=c("Sample","Species","Treatment"),variable.name="Family",value.name="Proportion")
sum_sig <- Rmisc::summarySE(sig_long, measurevar = "Proportion", groupvars = c("Treatment","Family"), na.rm=TRUE)
cols<-c("Untreated"="#E69F00","Control plots"="#56B4E9","Hypoxia plots"="#CC79A7")
sum_sig$Treatment<-factor(sum_sig$Treatment, levels=c("Untreated","Control plots","Hypoxia plots"))
#pdf("ANCOM_Families_Treatment_covarSpecies.pdf",width=8.5)
fams <- ggplot(sum_sig, aes(x=Family, y=Proportion))+
  geom_point(size=4,aes(color=Treatment))+
  scale_colour_manual(values=cols)+
  coord_flip()+
  theme_bw()+
  theme(axis.text.x=element_text(size=14))+
  theme(axis.text.y=element_text(size=14))+
  theme(axis.title.x=element_text(size=14))+
  theme(axis.title.y=element_text(size=14))+
  theme(legend.justification=c(1,1), legend.position=c(1,1))+
  geom_errorbar(aes(ymin=Proportion-se, ymax=Proportion+se), width=.1)+
  ggtitle("Siderastrea siderea")+
  theme(plot.title = element_text(face="italic"))+
  theme(legend.title = element_blank())+
  theme(legend.text = element_text(size=12))
fams
#dev.off()
```


```{r}
############ Agaracia lamarcki
# aggregate ASVs by family
dat <- tax_glom(ps_alam, taxrank = "Family")
#melt the data, so it's like a dataframe
datm <- psmelt(dat)
#Cast the new datatable with columns that are of interest
datc <- data.table::dcast(datm, Sample + Logger.MiniDOT + Treatment + Origin + Species ~ Family, value.var = 'Abundance', fun.aggregate = sum)
dim(datc) #dimensions of the table
#680 families
otud <- datc[,c(1,7:680)] #select the first column, and then all of the taxa columns  
colnames(otud)[1] <- "Sample.ID" #rename the first column to Sample.ID - this is to match ANCOM syntax
metadat <- sample_data(ps_alam) #get the sample data
metadat <- as.data.frame(as.matrix(metadat)) #make into into a matrix
# at this point, my sample id numbers are the row names, not a separate column, move row names to column with dplylr
metadat <- tibble::rownames_to_column(metadat, "Sample.ID") #make sure the sample names column is called Sample.ID
names(otud) <- make.names(names(otud)) #get the names from the table for 
otu_test <- otud #rename otud to otu_test, for syntax in ANCOM
metadat <- select(metadat, c("Sample.ID","Species","Treatment","Origin")) # use select to only use treatment columns of interest
map_test <- metadat #rename map_TEst
Vardat <- map_test #specify that this for Vardat - ANCOM syntax
#### ANCOM test - not adjusted, more than 2 levels = Kruskal Wallis
comparison_test_treat=ANCOM.main(OTUdat=otu_test, #calling the OTU table
                                 Vardat=map_test, #calling the metadata
                                 adjusted=FALSE, #true if covariates are to be included for adjustment
                                 repeated=FALSE, #repeated measure
                                 main.var="Treatment", #main variable or fator
                                 adj.formula= NULL, #other factors to include
                                 repeat.var=FALSE, #repeated measure
                                 long = FALSE, #longitudinal study
                                 multcorr=2,
                                 sig=0.05, #significance level
                                 prev.cut=0.90) #OTUs with proportion of zeroes greater than prev.cut are not included in the analysis
resA <- comparison_test_treat$W.taxa #taxa that significantly vary across factor level of interest
write.table(res,"ALAM_ANCOM_family_KruskallWallis_Treatment.txt",sep="\t",col.names=NA)
resA2 <- res[which(resA$detected_0.7==TRUE),] 
sig_sites <- glue::glue_collapse(droplevels(factor(resA2$otu.names)), sep = ", ") #this is to get a list of the families that are different
print(sig_sites)
##Family_XII, Marinifilaceae, Pseudomonadaceae, Nitrincolaceae, Marinobacteraceae

datc_relabund <-  sweep(datc[,7:680], 1, rowSums(datc[,7:680]), '/')
datc_relnames <- cbind(datc[,1:6],datc_relabund)
#only selet the significant families
sig_dis <- select(datc_relnames, Sample, Species, Treatment, Family_XII, Marinifilaceae, Pseudomonadaceae, Nitrincolaceae, Marinobacteraceae)
sig_long <- melt(sig_dis, id.vars=c("Sample","Species","Treatment"),variable.name="Family",value.name="Proportion")
library(Rmisc)
sum_sig <- Rmisc::summarySE(sig_long, measurevar = "Proportion", groupvars = c("Treatment","Family"), na.rm=TRUE)
cols<-c("Untreated"="#E69F00","Control plots"="#56B4E9","Hypoxia plots"="#CC79A7")
sum_sig$Treatment<-factor(sum_sig$Treatment, levels=c("Untreated","Control plots","Hypoxia plots"))
#pdf("ANCOM_Families_Treatment_covarSpecies.pdf",width=8.5)
fams <- ggplot(sum_sig, aes(x=Family, y=Proportion))+
  geom_point(size=4,aes(color=Treatment))+
  scale_colour_manual(values=cols)+
  coord_flip()+
  theme_bw()+
  theme(axis.text.x=element_text(size=14))+
  theme(axis.text.y=element_text(size=14))+
  theme(axis.title.x=element_text(size=14))+
  theme(axis.title.y=element_text(size=14))+
  theme(legend.justification=c(1,1), legend.position=c(1,1))+
  geom_errorbar(aes(ymin=Proportion-se, ymax=Proportion+se), width=.1)+
  ggtitle("Agaricia lamarcki")+
  theme(plot.title = element_text(face="italic"))+
  theme(legend.title = element_blank())+
  theme(legend.text = element_text(size=12))
fams
#dev.off()
```


```{r}
############ BOTH SPECIES TOGETHER
# aggregate ASVs by family
dat <- tax_glom(ps, taxrank = "Family")
#melt the data, so it's like a dataframe
datm <- psmelt(dat)
#Cast the new datatable with columns that are of interest
datc <- data.table::dcast(datm, Sample + Logger.MiniDOT + Treatment + Origin + Species ~ Family, value.var = 'Abundance', fun.aggregate = sum)
dim(datc) #dimensions of the table
#680 families
otud <- datc[,c(1,7:680)] #select the first column, and then all of the taxa columns  
colnames(otud)[1] <- "Sample.ID" #rename the first column to Sample.ID - this is to match ANCOM syntax
metadat <- sample_data(ps) #get the sample data
metadat <- as.data.frame(as.matrix(metadat)) #make into into a matrix
# at this point, my sample id numbers are the row names, not a separate column, move row names to column with dplylr
metadat <- tibble::rownames_to_column(metadat, "Sample.ID") #make sure the sample names column is called Sample.ID
names(otud) <- make.names(names(otud)) #get the names from the table for 
otu_test <- otud #rename otud to otu_test, for syntax in ANCOM
metadat <- select(metadat, c("Sample.ID","Species","Treatment","Origin")) # use select to only use treatment columns of interest
map_test <- metadat #rename map_TEst
Vardat <- map_test #specify that this for Vardat - ANCOM syntax
#### ANCOM test - not adjusted, more than 2 levels = Kruskal Wallis
comparison_test_treat=ANCOM.main(OTUdat=otu_test, #calling the OTU table
                                 Vardat=map_test, #calling the metadata
                                 adjusted=FALSE, #true if covariates are to be included for adjustment
                                 repeated=FALSE, #repeated measure
                                 main.var="Treatment", #main variable or fator
                                 adj.formula= NULL, #other factors to include
                                 repeat.var=FALSE, #repeated measure
                                 long = FALSE, #longitudinal study
                                 multcorr=2,
                                 sig=0.05, #significance level
                                 prev.cut=0.90) #OTUs with proportion of zeroes greater than prev.cut are not included in the analysis
resALL <- comparison_test_treat$W.taxa #taxa that significantly vary across factor level of interest
write.table(res,"ANCOM_family_KruskallWallis_Treatment.txt",sep="\t",col.names=NA)
resALL2 <- res[which(resALL$detected_0.7==TRUE),] 
#### ANCOM test - Adjusted by Coral species, ANOVA
comparison_test_treat=ANCOM.main(OTUdat=otu_test, #calling the OTU table
                                 Vardat=map_test, #calling the metadata
                                 adjusted=TRUE, #true if covariates are to be included for adjustment
                                 repeated=FALSE, #repeated measure
                                 main.var="Treatment", #main variable or fator
                                 adj.formula= "Species", #other factors to include
                                 repeat.var=FALSE, #repeated measure
                                 long = FALSE, #longitudinal study
                                 multcorr=2,
                                 sig=0.05, #significance level
                                 prev.cut=0.90) #OTUs with proportion of zeroes greater than prev.cut are not included in the analysis
resALL3 <- comparison_test_treat$W.taxa #taxa that significantly vary across factor level of interest
write.table(resALL3,"SSID_ANCOM_family_ANOVA_Treatment_covarSpecies.txt",sep="\t",col.names=NA)
resALL4 <- res[which(resALL3$detected_0.9==TRUE),] 
write.table(resALL4,"ANCOM_family_ANOVA_Treatment_covarSpecies.txt",sep="\t",col.names=NA)
sig_sites <- glue::glue_collapse(droplevels(factor(resALL4$otu.names)), sep = ", ") #this is to get a list of the families that are different
print(sig_sites)
#Family_XII, Marinifilaceae, Parvibaculaceae, Pseudomonadaceae, Marinobacteraceae, Nitrincolaceae, Francisellaceae, Shewanellaceae, Halomonadaceae, Midichloriaceae, Desulfovibrionaceae, Cellvibrionales

#Calculate relative abundance
datc_relabund <-  sweep(datc[,7:680], 1, rowSums(datc[,7:680]), '/')
datc_relnames <- cbind(datc[,1:6],datc_relabund)
#only selet the significant families
sig_dis <- select(datc_relnames, Sample, Species, Treatment, Family_XII, Marinifilaceae, Parvibaculaceae, Pseudomonadaceae, Marinobacteraceae, Nitrincolaceae, Francisellaceae, Shewanellaceae, Halomonadaceae, Midichloriaceae, Desulfovibrionaceae, Cellvibrionales)
sig_long <- melt(sig_dis,id.vars=c("Sample","Species","Treatment"),variable.name="Family",value.name="Proportion")
sum_sig <- Rmisc::summarySE(sig_long, measurevar = "Proportion", groupvars = c("Treatment","Family", "Species"), na.rm=TRUE)

write.table(sum_sig, "ANCOM_sum_sig_table.txt",sep="\t",col.names=NA)

#position=position_jitter(height=.00001)
#pdf("ANCOM_Families_Treatment_covarSpecies.pdf",width=11)
cols<-c("Control plots"="#E69F00","Hypoxia plots"="#CC79A7")
sum_sig$Treatment<-factor(sum_sig$Treatment, levels=c("Control plots","Hypoxia plots"))
fams <- ggplot(sum_sig, aes(x=Family, y=Proportion))+
  geom_point(size=4,aes(color=Treatment))+
  scale_colour_manual(values=cols)+
  facet_grid(.~Species)+
  coord_flip()+
  theme_bw()+
  theme(plot.margin=unit(c(1,1,1,1),"cm"))+
  theme(axis.text.x=element_text(size=10))+
  theme(axis.text.y=element_text(size=14))+
  theme(axis.title.x=element_text(size=14))+
  theme(axis.title.y=element_blank())+
  theme(strip.text.x = element_text(size=14, face="italic"))+
  theme(legend.justification=c(1,1), legend.position=c(1,1))+
  geom_errorbar(aes(ymin=Proportion-se, ymax=Proportion+se), width=.1)+
  theme(legend.title = element_blank())+
  theme(legend.text = element_text(size=12))
fams
#dev.off()
```


```{r}
#####ANCOM results for Logger 3 data (tent that decreased O2 drastically)
otu <- read.table("Panama_silva_nochloronomito_otu_table.txt",sep="\t",header=TRUE, row.names=1)
taxon <- read.table("Panama_silva_nochloronomito_taxa_table.txt",sep="\t",header=TRUE,row.names=1)
samples<-read.table("Panama_metadata.txt",sep="\t",header=T,row.names=1)
OTU = otu_table(otu, taxa_are_rows=FALSE)
taxon<-as.matrix(taxon)
TAX = tax_table(taxon)
sampledata = sample_data(samples)
ps <- phyloseq(otu_table(otu, taxa_are_rows=FALSE), 
               sample_data(samples), 
               tax_table(taxon))
ps
#16381 taxa and 57 samples
# remove control samples for plotting, remaining samples = 56
ps = subset_samples(ps, Species != "control")
ps

#remove unmanipulated, resident corals
ps = subset_samples(ps, Site == "Punta Caracol")
ps


#22823 taxa and 56 samples
#Separate coral species
ps_ssid <- subset_samples(ps, Species == "Siderastrea siderea")
ps_alam <- subset_samples(ps, Species == "Agaricia lamarcki")


############Siderastrea siderea
# aggregate ASVs by family
dat <- tax_glom(ps_ssid, taxrank = "Family")
#melt the data, so it's like a dataframe
datm <- psmelt(dat)
#Cast the new datatable with columns that are of interest
datc <- data.table::dcast(datm, Sample + Logger.MiniDOT + Origin + Species ~ Family, value.var = 'Abundance', fun.aggregate = sum)
dim(datc) #dimensions of the table
#679 families
otud <- datc[,c(1,7:679)] #select the first column, and then all of the taxa columns  
colnames(otud)[1] <- "Sample.ID" #rename the first column to Sample.ID - this is to match ANCOM syntax
metadat <- sample_data(ps_ssid) #get the sample data
metadat <- as.data.frame(as.matrix(metadat)) #make into into a matrix
# at this point, my sample id numbers are the row names, not a separate column, move row names to column with dplylr
metadat <- tibble::rownames_to_column(metadat, "Sample.ID") #make sure the sample names column is called Sample.ID
names(otud) <- make.names(names(otud)) #get the names from the table for 
otu_test <- otud #rename otud to otu_test, for syntax in ANCOM
metadat <- select(metadat, c("Sample.ID","Species","Logger.MiniDOT")) # use select to only use Logger.MiniDOT columns of interest
map_test <- metadat #rename map_TEst
Vardat <- map_test #specify that this for Vardat - ANCOM syntax
#### ANCOM test - not adjusted, more than 2 levels = Kruskal Wallis
comparison_test_log=ANCOM.main(OTUdat=otu_test, #calling the OTU table
                                 Vardat=map_test, #calling the metadata
                                 adjusted=FALSE, #true if covariates are to be included for adjustment
                                 repeated=FALSE, #repeated measure
                                 main.var="Logger.MiniDOT", #main variable or fator
                                 adj.formula= NULL, #other factors to include
                                 repeat.var=FALSE, #repeated measure
                                 long = FALSE, #longitudinal study
                                 multcorr=2,
                                 sig=0.05, #significance level
                                 prev.cut=0.90) #OTUs with proportion of zeroes greater than prev.cut are not included in the analysis
res <- comparison_test_treat$W.taxa #taxa that significantly vary across factor level of interest
write.table(res,"SSID_ANCOM_family_KruskallWallis_Logger.MiniDOT.txt",sep="\t",col.names=NA)
res2 <- res[which(res$detected_0.7==TRUE),] 
sig_sites <- glue::glue_collapse(droplevels(factor(res2$otu.names)), sep = ", ") #this is to get a list of the families that are different
print(sig_sites)
#Family_XII, JTB215, Marinobacteraceae, Pseudomonadaceae, Halomonadaceae, Acholeplasmataceae, Shewanellaceae, Nitrincolaceae, Rhodospirillales, Arcobacteraceae, Marinifilaceae, Clostridiaceae_1, Izimaplasmatales, Marinilabiliaceae, Prolixibacteraceae, GoM.GC232.4463.Bac1, Mollicutes_RF39, Bacteroidales, Desulfovibrionaceae, MSBL8, Francisellaceae, Family_XIII, Spirochaetaceae, Clostridiaceae, Clostridiales, Desulfobacteraceae, Clostridiaceae_4, Lachnospiraceae, Firmicutes, Midichloriaceae, Peptostreptococcaceae, Bacteroidetes_BD2.2, Fusobacteriaceae, Izimaplasmataceae, Desulfobulbaceae, OPB56, Defluviitaleaceae, Melioribacteraceae, P.palmC41, Vibrionaceae, Colwelliaceae, Desulfuromonadaceae, Micrococcaceae, Absconditabacteriales_.SR1., LCP.89, Puniceispirillales, Marinomonadaceae, Gammaproteobacteria, Holosporaceae, Thiotrichaceae, Litoricolaceae, Flammeovirgaceae, Clade_III, Kangiellaceae, Cryomorphaceae, Sphingomonadaceae, Proteobacteria, Sneathiellaceae, BIrii41, Parvibaculaceae, OCS116_clade, Thermodesulfovibrionia, Clade_II, Cellvibrionaceae, KI89A_clade, Erysipelotrichaceae, Syntrophobacteraceae, Hyphomonadaceae, AEGEAN.169_marine_group, Arenicellaceae, Deltaproteobacteria, UBA10353_marine_group, Spongiibacteraceae, Bacteroidia, Idiomarinaceae, Thermoanaerobaculaceae, OM182_clade
#Calculate relative abundance
datc_relabund <-  sweep(datc[,7:679], 1, rowSums(datc[,7:679]), '/')
datc_relnames <- cbind(datc[,1:6],datc_relabund)
#only selet the significant families
sig_dis <- select(datc_relnames, Sample, Species, Logger.MiniDOT, Family_XII, JTB215, Marinobacteraceae, Pseudomonadaceae, Halomonadaceae, Acholeplasmataceae, Shewanellaceae, Nitrincolaceae, Rhodospirillales, Arcobacteraceae, Marinifilaceae, Clostridiaceae_1, Izimaplasmatales, Marinilabiliaceae, Prolixibacteraceae, Mollicutes_RF39, Bacteroidales, Desulfovibrionaceae, MSBL8, Francisellaceae, Family_XIII, Spirochaetaceae, Clostridiaceae, Clostridiales, Desulfobacteraceae, Clostridiaceae_4, Lachnospiraceae, Firmicutes, Midichloriaceae, Peptostreptococcaceae, "Bacteroidetes_BD2-2", Fusobacteriaceae, Izimaplasmataceae, Desulfobulbaceae, OPB56, Defluviitaleaceae, Melioribacteraceae, P.palmC41, Vibrionaceae, Colwelliaceae, Desulfuromonadaceae, Micrococcaceae, "Absconditabacteriales_(SR1)", "LCP-89", Puniceispirillales, Marinomonadaceae, Gammaproteobacteria, Holosporaceae, Thiotrichaceae, Litoricolaceae, Flammeovirgaceae, Clade_III, Kangiellaceae, Cryomorphaceae, Sphingomonadaceae, Proteobacteria, Sneathiellaceae, BIrii41, Parvibaculaceae, OCS116_clade, Thermodesulfovibrionia, Clade_II, Cellvibrionaceae, KI89A_clade, Erysipelotrichaceae, Syntrophobacteraceae, Hyphomonadaceae, "AEGEAN-169_marine_group", Arenicellaceae, Deltaproteobacteria, UBA10353_marine_group, Spongiibacteraceae, Bacteroidia, Idiomarinaceae, Thermoanaerobaculaceae, OM182_clade)
sig_long <- melt(sig_dis, id.vars=c("Sample","Species","Logger.MiniDOT"),variable.name="Family",value.name="Proportion")
sum_sig <- Rmisc::summarySE(sig_long, measurevar = "Proportion", groupvars = c("Logger.MiniDOT","Family"), na.rm=TRUE)
cols<-c("1"="#C9A456","2"="#E27741","3"="#1A5276","4"="#CB4E07","5"="#8A9B5E", "8"="#B04029", "9"="#ABC9CB")
sum_sig$Logger.MiniDOT<-factor(sum_sig$Logger.MiniDOT, levels=c("1","2","3","4","5","8","9"))
#pdf("ANCOM_Families_Logger.MiniDOT_covarSpecies.pdf",width=8.5)
fams <- ggplot(sum_sig, aes(x=Family, y=Proportion))+
  geom_point(size=4,aes(color=Logger.MiniDOT))+
  scale_colour_manual(values=cols)+
  coord_flip()+
  theme_bw()+
  theme(axis.text.x=element_text(size=14))+
  theme(axis.text.y=element_text(size=14))+
  theme(axis.title.x=element_text(size=14))+
  theme(axis.title.y=element_text(size=14))+
  theme(legend.justification=c(1,1), legend.position=c(1,1))+
  geom_errorbar(aes(ymin=Proportion-se, ymax=Proportion+se), width=.1)+
  ggtitle("Siderastrea siderea")+
  theme(plot.title = element_text(face="italic"))+
  theme(legend.title = element_blank())+
  theme(legend.text = element_text(size=12))
fams
#dev.off()
```

```{r}
############ Agaracia lamarcki
# aggregate ASVs by family
dat <- tax_glom(ps_alam, taxrank = "Family")
#melt the data, so it's like a dataframe
datm <- psmelt(dat)
#Cast the new datatable with columns that are of interest
datc <- data.table::dcast(datm, Sample + Logger.MiniDOT + Treatment + Origin + Species ~ Family, value.var = 'Abundance', fun.aggregate = sum)
dim(datc) #dimensions of the table
#680 families
otud <- datc[,c(1,7:680)] #select the first column, and then all of the taxa columns  
colnames(otud)[1] <- "Sample.ID" #rename the first column to Sample.ID - this is to match ANCOM syntax
metadat <- sample_data(ps_alam) #get the sample data
metadat <- as.data.frame(as.matrix(metadat)) #make into into a matrix
# at this point, my sample id numbers are the row names, not a separate column, move row names to column with dplylr
metadat <- tibble::rownames_to_column(metadat, "Sample.ID") #make sure the sample names column is called Sample.ID
names(otud) <- make.names(names(otud)) #get the names from the table for 
otu_test <- otud #rename otud to otu_test, for syntax in ANCOM
metadat <- select(metadat, c("Sample.ID","Species","Logger.MiniDOT")) # use select to only use Logger.MiniDOT columns of interest
map_test <- metadat #rename map_TEst
Vardat <- map_test #specify that this for Vardat - ANCOM syntax
#### ANCOM test - not adjusted, more than 2 levels = Kruskal Wallis
comparison_test_log=ANCOM.main(OTUdat=otu_test, #calling the OTU table
                                 Vardat=map_test, #calling the metadata
                                 adjusted=FALSE, #true if covariates are to be included for adjustment
                                 repeated=FALSE, #repeated measure
                                 main.var="Logger.MiniDOT", #main variable or fator
                                 adj.formula= NULL, #other factors to include
                                 repeat.var=FALSE, #repeated measure
                                 long = FALSE, #longitudinal study
                                 multcorr=2,
                                 sig=0.05, #significance level
                                 prev.cut=0.90) #OTUs with proportion of zeroes greater than prev.cut are not included in the analysis
resA <- comparison_test_treat$W.taxa #taxa that significantly vary across factor level of interest
write.table(res,"ALAM_ANCOM_family_KruskallWallis_Treatment.txt",sep="\t",col.names=NA)
resA2 <- res[which(resA$detected_0.7==TRUE),] 
sig_sites <- glue::glue_collapse(droplevels(factor(resA2$otu.names)), sep = ", ") #this is to get a list of the families that are different
print(sig_sites)
#Family_XII, JTB215, Marinobacteraceae, Pseudomonadaceae, Halomonadaceae, Acholeplasmataceae, Shewanellaceae, Nitrincolaceae, Rhodospirillales, Arcobacteraceae, Marinifilaceae, Clostridiaceae_1, Izimaplasmatales, Marinilabiliaceae, Prolixibacteraceae, GoM.GC232.4463.Bac1, Mollicutes_RF39, Bacteroidales, Desulfovibrionaceae, MSBL8, Francisellaceae, Family_XIII, Spirochaetaceae, Clostridiaceae, Clostridiales, Desulfobacteraceae, Clostridiaceae_4, Lachnospiraceae, Firmicutes, Midichloriaceae, Peptostreptococcaceae, Bacteroidetes_BD2.2, Fusobacteriaceae, Izimaplasmataceae, Desulfobulbaceae, OPB56, Defluviitaleaceae, Melioribacteraceae, P.palmC41, Vibrionaceae, Colwelliaceae, Desulfuromonadaceae, Micrococcaceae, Absconditabacteriales_.SR1., LCP.89, Puniceispirillales, Marinomonadaceae, Gammaproteobacteria, Holosporaceae, Thiotrichaceae, Litoricolaceae, Flammeovirgaceae, Clade_III, Kangiellaceae, Cryomorphaceae, Sphingomonadaceae, Proteobacteria, Sneathiellaceae, BIrii41, Parvibaculaceae, OCS116_clade, Thermodesulfovibrionia, Clade_II, Cellvibrionaceae, KI89A_clade, Erysipelotrichaceae, Syntrophobacteraceae, Hyphomonadaceae, AEGEAN.169_marine_group, Arenicellaceae, Deltaproteobacteria, UBA10353_marine_group, Spongiibacteraceae, Bacteroidia, Idiomarinaceae, Thermoanaerobaculaceae, OM182_clade

datc_relabund <-  sweep(datc[,7:680], 1, rowSums(datc[,7:680]), '/')
datc_relnames <- cbind(datc[,1:6],datc_relabund)
#only selet the significant families
sig_dis <- select(datc_relnames, Sample, Species, Logger.MiniDOT, Family_XII, JTB215, Marinobacteraceae, Pseudomonadaceae, Halomonadaceae, Acholeplasmataceae, Shewanellaceae, Nitrincolaceae, Rhodospirillales, Arcobacteraceae, Marinifilaceae, Clostridiaceae_1, Izimaplasmatales, Marinilabiliaceae, Prolixibacteraceae, Mollicutes_RF39, Bacteroidales, Desulfovibrionaceae, MSBL8, Francisellaceae, Family_XIII, Spirochaetaceae, Clostridiaceae, Clostridiales, Desulfobacteraceae, Clostridiaceae_4, Lachnospiraceae, Firmicutes, Midichloriaceae, Peptostreptococcaceae, "Bacteroidetes_BD2-2", Fusobacteriaceae, Izimaplasmataceae, Desulfobulbaceae, OPB56, Defluviitaleaceae, Melioribacteraceae, P.palmC41, Vibrionaceae, Colwelliaceae, Desulfuromonadaceae, Micrococcaceae, "Absconditabacteriales_(SR1)", "LCP-89", Puniceispirillales, Marinomonadaceae, Gammaproteobacteria, Holosporaceae, Thiotrichaceae, Litoricolaceae, Flammeovirgaceae, Clade_III, Kangiellaceae, Cryomorphaceae, Sphingomonadaceae, Proteobacteria, Sneathiellaceae, BIrii41, Parvibaculaceae, OCS116_clade, Thermodesulfovibrionia, Clade_II, Cellvibrionaceae, KI89A_clade, Erysipelotrichaceae, Syntrophobacteraceae, Hyphomonadaceae, "AEGEAN-169_marine_group", Arenicellaceae, Deltaproteobacteria, UBA10353_marine_group, Spongiibacteraceae, Bacteroidia, Idiomarinaceae, Thermoanaerobaculaceae, OM182_clade)
sig_long <- melt(sig_dis, id.vars=c("Sample","Species","Logger.MiniDOT"),variable.name="Family",value.name="Proportion")
library(Rmisc)
sum_sig <- Rmisc::summarySE(sig_long, measurevar = "Proportion", groupvars = c("Logger.MiniDOT","Family"), na.rm=TRUE)
cols<-c("1"="#C9A456","2"="#E27741","3"="#1A5276","4"="#CB4E07","5"="#8A9B5E","8"="#B04029", "9"="#ABC9CB")
sum_sig$Logger.MiniDOT<-factor(sum_sig$Logger.MiniDOT, levels=c("1","2","3","4","5","8","9"))
#pdf("ANCOM_Families_Logger.MiniDOT_covarSpecies.pdf",width=8.5)
fams <- ggplot(sum_sig, aes(x=Family, y=Proportion))+
  geom_point(size=4,aes(color=Logger.MiniDOT))+
  scale_colour_manual(values=cols)+
  coord_flip()+
  theme_bw()+
  theme(axis.text.x=element_text(size=14))+
  theme(axis.text.y=element_text(size=14))+
  theme(axis.title.x=element_text(size=14))+
  theme(axis.title.y=element_text(size=14))+
  theme(legend.justification=c(1,1), legend.position=c(1,1))+
  geom_errorbar(aes(ymin=Proportion-se, ymax=Proportion+se), width=.1)+
  ggtitle("Agaricia lamarcki")+
  theme(plot.title = element_text(face="italic"))+
  theme(legend.title = element_blank())+
  theme(legend.text = element_text(size=12))
fams
#dev.off()
```


```{r}
############ BOTH SPECIES TOGETHER
# aggregate ASVs by family
dat <- tax_glom(ps, taxrank = "Family")
#melt the data, so it's like a dataframe
datm <- psmelt(dat)
#Cast the new datatable with columns that are of interest
datc <- data.table::dcast(datm, Sample + Logger.MiniDOT + Treatment + Origin + Species ~ Family, value.var = 'Abundance', fun.aggregate = sum)
dim(datc) #dimensions of the table
#680 families
otud <- datc[,c(1,7:680)] #select the first column, and then all of the taxa columns  
colnames(otud)[1] <- "Sample.ID" #rename the first column to Sample.ID - this is to match ANCOM syntax
metadat <- sample_data(ps) #get the sample data
metadat <- as.data.frame(as.matrix(metadat)) #make into into a matrix
# at this point, my sample id numbers are the row names, not a separate column, move row names to column with dplylr
metadat <- tibble::rownames_to_column(metadat, "Sample.ID") #make sure the sample names column is called Sample.ID
names(otud) <- make.names(names(otud)) #get the names from the table for 
otu_test <- otud #rename otud to otu_test, for syntax in ANCOM
metadat <- select(metadat, c("Sample.ID","Species","Logger.MiniDOT")) # use select to only use logger columns of interest
map_test <- metadat #rename map_TEst
Vardat <- map_test #specify that this for Vardat - ANCOM syntax
#### ANCOM test - not adjusted, more than 2 levels = Kruskal Wallis
comparison_test_log=ANCOM.main(OTUdat=otu_test, #calling the OTU table
                                 Vardat=map_test, #calling the metadata
                                 adjusted=FALSE, #true if covariates are to be included for adjustment
                                 repeated=FALSE, #repeated measure
                                 main.var="Logger.MiniDOT", #main variable or fator
                                 adj.formula= NULL, #other factors to include
                                 repeat.var=FALSE, #repeated measure
                                 long = FALSE, #longitudinal study
                                 multcorr=2,
                                 sig=0.05, #significance level
                                 prev.cut=0.90) #OTUs with proportion of zeroes greater than prev.cut are not included in the analysis
resALL <- comparison_test_treat$W.taxa #taxa that significantly vary across factor level of interest
write.table(res,"ANCOM_family_KruskallWallis_Treatment.txt",sep="\t",col.names=NA)
resALL2 <- res[which(resALL$detected_0.7==TRUE),] 
#### ANCOM test - Adjusted by Coral species, ANOVA
comparison_test_treat=ANCOM.main(OTUdat=otu_test, #calling the OTU table
                                 Vardat=map_test, #calling the metadata
                                 adjusted=TRUE, #true if covariates are to be included for adjustment
                                 repeated=FALSE, #repeated measure
                                 main.var="Logger.MiniDOT", #main variable or fator
                                 adj.formula= "Species", #other factors to include
                                 repeat.var=FALSE, #repeated measure
                                 long = FALSE, #longitudinal study
                                 multcorr=2,
                                 sig=0.05, #significance level
                                 prev.cut=0.90) #OTUs with proportion of zeroes greater than prev.cut are not included in the analysis
resALL3 <- comparison_test_treat$W.taxa #taxa that significantly vary across factor level of interest
write.table(resALL3,"ANCOM_family_ANOVA_Logger_covarSpecies.txt",sep="\t",col.names=NA)
resALL4 <- res[which(resALL3$detected_0.9==TRUE),] 
write.table(resALL4,"resALL4.txt",sep="\t",col.names=NA)
sig_sites <- glue::glue_collapse(droplevels(factor(resALL4$otu.names)), sep = ", ") #this is to get a list of the families that are different
print(sig_sites)
#Family_XII, JTB215, Marinobacteraceae, Pseudomonadaceae, Halomonadaceae, Acholeplasmataceae, Shewanellaceae, Nitrincolaceae, Rhodospirillales, Arcobacteraceae, Marinifilaceae, Clostridiaceae_1, Izimaplasmatales, Marinilabiliaceae, Prolixibacteraceae, GoM.GC232.4463.Bac1, Mollicutes_RF39, Bacteroidales, Desulfovibrionaceae, MSBL8, Francisellaceae, Family_XIII, Spirochaetaceae, Clostridiaceae, Clostridiales, Desulfobacteraceae, Clostridiaceae_4, Lachnospiraceae, Firmicutes, Midichloriaceae, Peptostreptococcaceae, Bacteroidetes_BD2.2, Fusobacteriaceae, Izimaplasmataceae, Desulfobulbaceae, OPB56, Defluviitaleaceae, Melioribacteraceae, P.palmC41, Vibrionaceae, Colwelliaceae

#Calculate relative abundance
datc_relabund <-  sweep(datc[,7:680], 1, rowSums(datc[,7:680]), '/')
datc_relnames <- cbind(datc[,1:6],datc_relabund)
#only selet the significant families
sig_dis <- select(datc_relnames, Sample, Species, Logger.MiniDOT, Family_XII, JTB215, Marinobacteraceae, Pseudomonadaceae, Halomonadaceae, Acholeplasmataceae, Shewanellaceae, Nitrincolaceae, Rhodospirillales, Arcobacteraceae, Marinifilaceae, Clostridiaceae_1, Izimaplasmatales, Marinilabiliaceae, Prolixibacteraceae, Mollicutes_RF39, Bacteroidales, Desulfovibrionaceae, MSBL8, Francisellaceae, Family_XIII, Spirochaetaceae, Clostridiaceae, Clostridiales, Desulfobacteraceae, Clostridiaceae_4, Lachnospiraceae, Firmicutes, Midichloriaceae, Peptostreptococcaceae, "Bacteroidetes_BD2-2", Fusobacteriaceae, Izimaplasmataceae, Desulfobulbaceae, OPB56, Defluviitaleaceae, Melioribacteraceae, P.palmC41, Vibrionaceae, Colwelliaceae)
sig_long <- melt(sig_dis, id.vars=c("Sample","Species","Logger.MiniDOT"),variable.name="Family",value.name="Proportion")
sum_sig <- Rmisc::summarySE(sig_long, measurevar = "Proportion", groupvars = c("Species","Logger.MiniDOT","Family"), na.rm=TRUE)

write.table(sum_sig, "ANCOM_sum_sig_Logger3_table.txt",sep="\t",col.names=NA)

#position=position_jitter(height=.00001)
#pdf("ANCOM_Families_Logger_covarSpecies.pdf",width=14,height=16)
cols<-c("1"="#C9A456","2"="#E27741","3"="#1A5276","4"="#CB4E07","5"="#8A9B5E","8"="#B04029","9"="#ABC9CB")
sum_sig$Logger.MiniDOT<-factor(sum_sig$Logger.MiniDOT, levels=c("1","2","3","4","5","8","9"))
fams <- ggplot(sum_sig, aes(x=Family, y=Proportion))+
  geom_point(size=4,aes(color=Logger.MiniDOT))+
  scale_colour_manual(values=cols)+
  facet_grid(.~Species)+
  coord_flip()+
  theme_bw()+
  theme(plot.margin=unit(c(1,1,1,1),"cm"))+
  theme(axis.text.x=element_text(size=10))+
  theme(axis.text.y=element_text(size=14))+
  theme(axis.title.x=element_text(size=14))+
  theme(axis.title.y=element_blank())+
  theme(strip.text.x = element_text(size=14, face="italic"))+
  theme(legend.justification=c(1,1), legend.position=c(1,1))+
  geom_errorbar(aes(ymin=Proportion-se, ymax=Proportion+se), width=.1)+
  theme(legend.title = element_blank())+
  theme(legend.text = element_text(size=12))
fams
#dev.off()
```


```{r}
#Only select families with the greatest difference in relative abundance
sig_dis <- select(datc_relnames, Sample, Species, Logger.MiniDOT,Family_XII, Desulfovibrionaceae, Arcobacteraceae)
sig_long <- melt(sig_dis, id.vars=c("Sample","Species","Logger.MiniDOT"),variable.name="Family",value.name="Proportion")
sum_sig <- Rmisc::summarySE(sig_long, measurevar = "Proportion", groupvars = c("Species","Logger.MiniDOT","Family"), na.rm=TRUE)
#position=position_jitter(height=.00001)
pdf("ANCOM_Families_Logger3Fanilies_covarSpecies.pdf",width=11)
cols<-c("1"="#C9A456","2"="#E27741","3"="#1A5276","4"="#CB4E07","5"="#8A9B5E","8"="#B04029","9"="#ABC9CB")
sum_sig$Logger.MiniDOT<-factor(sum_sig$Logger.MiniDOT, levels=c("1","2","3","4","5","8","9"))
fams <- ggplot(sum_sig, aes(x=Family, y=Proportion))+
  geom_point(size=4,aes(color=Logger.MiniDOT))+
  scale_colour_manual(values=cols)+
  facet_grid(.~Species)+
  coord_flip()+
  theme_bw()+
  theme(plot.margin=unit(c(1,1,1,1),"cm"))+
  theme(axis.text.x=element_text(size=10))+
  theme(axis.text.y=element_text(size=14))+
  theme(axis.title.x=element_text(size=14))+
  theme(axis.title.y=element_blank())+
  theme(strip.text.x = element_text(size=14, face="italic"))+
  theme(legend.justification=c(1,1), legend.position=c(1,1))+
  geom_errorbar(aes(ymin=Proportion-se, ymax=Proportion+se), width=.1)+
  theme(legend.title = element_blank())+
  theme(legend.text = element_text(size=12))
fams
dev.off()
```

## Indicator species analysis

# Add metadata columns to ASV table with relative abundances as shown at the top of this tutorial: https://jkzorz.github.io/2019/07/02/Indicator-species-analysis.html


```{r, echo=FALSE}
library(indicspecies)
asv <- read.table("Panama_ps_ra_silva_nochloronomito_otu_table_INDIC.txt",sep="\t",header=TRUE)

#subset for experimental corals only
asv_pc<-subset(asv, Site =='Punta Caracol')

# In the following line, change "6" to the column number that starts your ASV abundances
abund = asv_pc[,5:ncol(asv)]
treatment=asv_pc$Treatment
inv = multipatt(abund, treatment, func = "r.g", control = how(nperm=9999))
summary(inv)


species=asv_pc$Species
inv2 = multipatt(abund, species, func = "r.g", control = how(nperm=9999))
summary(inv2)

##Make a boxplot of treatment indic results
pc<-read.table("Panama_ps_ra_silva_nochloronomito_otu_table_INDIC.txt",sep="\t",header=TRUE)

#subset for experimental corals only
sub_puc<-subset(pc, Site =='Punta Caracol')

##Next, we subset the data as desired by calling the specific columns we want to include in our boxplot figure. I’ve chosen the top four most significant ASVs associated with the “Hypoxia plots” grouping (stat-value>.50)
sub_pc<-data.frame(Treatment = sub_puc$Treatment, "Alteromonas"= sub_puc$ASV_17, "Neptuniibacter"= sub_puc$ASV_256, "Aestuariicella" = sub_puc$ASV_314, "Marinobacter"= sub_puc$ASV_425)

##Next, we convert the data frame from a “wide” format to a “long” format 
sub_pc_m = melt(sub_pc, id = "Treatment")

##Now we can plot the data using geom_boxplot from ggplot2
sub_pc_m$Treatment<-factor(sub_pc_m$Treatment, levels=c("Control plots","Hypoxia plots"))
cols<-c("Control plots"="#E69F00","Hypoxia plots"="#CC79A7")
pdf("INDIC_ASVs_Treatment_boxplot.pdf",width=8)
gg<-ggplot(sub_pc_m, aes(x = variable, y = value)) + 
    geom_boxplot(outlier.shape=NA)+
    geom_point(aes(color=Treatment))+
    geom_jitter(position=position_jitter(width=.1, height=0),aes(color=Treatment))+
     theme(legend.title = element_text(size = 12, face = "bold"), 
     legend.text = element_text(size = 10, face = "bold"), legend.position = "right", 
     axis.text.x = element_text(face="bold.italic",colour = "black", size = 12), 
     axis.text.y = element_text(face = "bold", size = 12, colour = "black"), 
     axis.title.y = element_text(face = "bold", size = 14, colour = "black"), 
     panel.background = element_blank(), 
     panel.border = element_rect(fill = NA, colour = "black"), 
     legend.key=element_blank()) + 
     labs(x= "", y = "Proportion") + 
     scale_color_manual(values = cols)
gg
dev.off()
```