# Shifts in the coral microbiome in response to in situ experimental deoxygenation
This project involves the analysis of microbiomes of coral fragments exposed to in situ marine hypoxia in Bahía Almirante, Panama. The sequencing reads with primers and adaptors removed by cutadapt are available in NCBI’s Sequence Read Archive under Bioproject PRJNA641080. This repository includes the R script to recreate the figures in (Howard et al submitted to...), and the original metadata, ASV, and taxonomy tables from our analysis. The file of session information provides all of the versions of the packages used in R to complete the analysis in (Howard et al 2021). The provided ASV and taxonomy tables are the original files provided by the dada2 analysis. The script includes using phyloseq to remove chloroplast and mitochondrial reads, as well as using R to fill in empty or "NA" cells in the taxonomy with the lowest taxonomy assignment available. Please note that if the dada2 analysis is repeated, it will result in a slightly different ASV table every time, although the overall interpretations will remain the same. Analysis includes determining the amplicon sequence variants (ASVs) with dada2, assigning taxonomy with the Silva database v.132, and community composition analysis with a combination of phyloseq, vegan, and CoDaSeq tools. The ANCOM function is used in R for community analysis (not available as an R package).
